models/resnet/resnet.hpp

```diff
@@ -0,0 +1,488 @@
+   * @param padH Padding height of the input.
+   */
+  template <typename AddMergeType = ann::AddMerge<>>
+  void DownSample(AddMergeType* resBlock,
+                  const size_t inSize,
+                  const size_t outSize,
+                  const size_t downSampleInputWidth,
+                  const size_t downSampleInputHeight,
+                  const size_t strideWidth = 2,
+                  const size_t strideHeight = 2)
+  {
+    mlpack::Log::Info << "DownSample (" << std::endl;
+    ann::Sequential<>* downSampleBlock = new ann::Sequential<>();
+    ConvolutionBlock(downSampleBlock, inSize, outSize, strideWidth,
+        strideHeight, 1, 1, 0, 0, true, downSampleInputWidth,
+        downSampleInputHeight);
+    mlpack::Log::Info << ")" <<std::endl;
+
+    resBlock->Add(downSampleBlock);
+  }
+
+
+  /**
+   * Adds a ReLU Layer.
+   *
+   * @tparam SequentialType Layer type in which ReLU layer will be added.
+   */
+  template<typename SequentialType = ann::Sequential<>>
+  void ReLULayer(SequentialType* baseLayer)
+  {
+    baseLayer->Add(new ann::ReLULayer<>);
+    mlpack::Log::Info << "Relu" << std::endl;
+  }
+
+  /**
+   * Adds basicBlock block for ResNet 18 and 34.
+   *
+   * It's represented as:
+   * 
+   * \code
+   * resBlock - AddMerge layer
+   * {
+   *   sequentialBlock - sequentialLayer
+   *   {
+   *     ConvolutionBlock(inSize, outSize, strideWidth, strideHeight)
+   *     ReLU
+   *     ConvolutionBlock(inSize, outSize)
+   *   }
+   *
+   *   sequentialLayer
+   *   {
+   *     if downsample == true
+   *       ConvolutionBlock(inSize, outSize, downSampleInputWidth,
+   *           downSampleInputHeight)
+   *     else
+   *       IdentityLayer
+   *   }
+   * 
+   *   ReLU
+   * }
+   * \endcode
+   * 
+   * @param inSize Number of input maps.
+   * @param outSize Number of output maps.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param downSample If there will be a downSample block or not, default
+   *     false.
+   */
+  void BasicBlock(const size_t inSize,
+                  const size_t outSize,
+                  const size_t strideWidth = 1,
+                  const size_t strideHeight = 1,
+                  const bool downSample = false)
+  {
+    downSampleInputWidth = inputWidth;
+    downSampleInputHeight = inputHeight;
+
+    ann::Sequential<>* basicBlock = new ann::Sequential<>();
+    ann::AddMerge<>* resBlock = new ann::AddMerge<>(true, true);
+    ann::Sequential<>* sequentialBlock = new ann::Sequential<>();
+    ConvolutionBlock(sequentialBlock, inSize, outSize, strideWidth,
+        strideHeight);
+    ReLULayer(sequentialBlock);
+    ConvolutionBlock(sequentialBlock, outSize, outSize);
+
+    resBlock->Add(sequentialBlock);
+
+    if (downSample == true)
+    {
+      DownSample(resBlock, inSize, outSize, downSampleInputWidth,
+          downSampleInputHeight);
+    }
```

Comment:
```suggestion

    if (downSample == true)
    {
      ann::Sequential<>* downSampleBlock = new ann::Sequential<>();
       ConvolutionBlock(downSampleBlock, inSize, outSize, strideWidth,
           strideHeight, 1, 1, 0, 0, true, downSampleInputWidth,
           downSampleInputHeight);
      resBlock->Add(downSampleBlock);
    }
```

and the function for convolution block can be changed to :
 ```
void ConvolutionBlock( ann::Sequential<>* baseLayer,
                         const size_t inSize,
                         const size_t outSize,
                         const size_t strideWidth = 1,
                         const size_t strideHeight = 1,
                         const size_t kernelWidth = 3,
                         const size_t kernelHeight = 3,
                         const size_t padW = 1,
                         const size_t padH = 1,
                         const bool downSample = false,
                         const size_t downSampleInputWidth = 0,
                         const size_t downSampleInputHeight = 0)
{

     if (downSample)
     {
       mlpack::Log::Info << "Downsample ";
       inputWidth = downSampleInputWidth;
       inputHeight = downSampleInputHeight;
     }

     baseLayer->Add(new ann::Convolution<>(inSize, outSize, kernelWidth,
         kernelHeight, strideWidth, strideHeight, padW, padH, convInputWidth,
         convInputHeight));
     mlpack::Log::Info << "Convolution: " << "(" << inSize << ", " <<
         convInputWidth << ", " << convInputHeight << ")" << " ---> (";


    // Updating input dimensions.
    inputWidth = ConvOutSize(convInputWidth, kernelWidth, strideWidth, padW);
    inputHeight = ConvOutSize(convInputHeight, kernelHeight, strideHeight,
           padH);


     mlpack::Log::Info << outSize << ", " << convInputWidth << ", " <<
         convInputHeight << ")" << std::endl;

     baseLayer->Add(new ann::BatchNorm<>(outSize, 1e-5));
     mlpack::Log::Info << "BatchNorm: " << "(" << outSize << ")" << " ---> ("
         << outSize << ")" << std::endl;
}
```

This solves the comment I left about input width / height above. Also we can remove template if it is always taking sequential layer as input.

Reply:
Hey, there is a problem with this, doing this changes the vars `inputWidth` and `inputHeight` which keep a track of input dimensions and changing them when a downsample block occurs messes up with the order they should go in, that's why in the current code i create two temp variables, Also do you think we should replace the downsaple block at two places. right the function itself takes around 15-18 lines, removing that and adding conv and seq layers at two places is again going to create the same number of LOC.

Reply:
Lets have a call at 6 PM IST if possible.

Reply:
sure, will be there.
Would meet be okay ?

Reply:
@kartikdutt18 sent the inivtes.
@zoq sent you an invite too, feel free to drop by if you want to.